src/sage/schemes/hodge_special_fourfolds/sff.py

```diff
@@ -0,0 +1,3099 @@
+            t = PolynomialRing(RationalField(), 1, 't').gen()
+            for i in range(self.dimension() - 1):
+                P = P(t)-P(t-1)
+            self._sectional_genus = Integer(1-P(0))
+            return self._sectional_genus
+
+    def topological_euler_characteristic(self, verbose=None, UseMacaulay2=False):
+        r"""Return the topological Euler characteristic of the variety
+
+        .. WARNING::
+
+            This uses a probabilistic approach which could give wrong answers
+            (especially over finite fields of small order).
+
+        With the input ``UseMacaulay2=True`` the computation is transferred to ``Macaulay2``.
+
+        OUTPUT:
+
+        An integer.
+
+        EXAMPLES::
+
+            sage: X = Veronese(2,2); X
+            surface of degree 4 and sectional genus 0 in PP^5 cut out by 6 hypersurfaces of degree 2
+            sage: X.topological_euler_characteristic()
+            3
+
+        """
+        if verbose is None:
+            verbose = __VERBOSE__
+        if not(isinstance(UseMacaulay2,bool) and isinstance(verbose,bool)):
+            raise TypeError("expected True or False")
+        try:
+            return self._topological_euler_characteristic
+        except AttributeError:
+            if UseMacaulay2 or min(self.degrees_generators()) != max(self.degrees_generators()):
+                if verbose:
+                    print("--topological_euler_characteristic(): transferring computation to Macaulay2...")
+                return self._topological_Euler_characteristic_macaulay2(verbose=verbose)
+            f = rational_map(self)
+            m = f._degree_forms()
+            d = f.projective_degrees()
+            r = self.dimension()
+            n = self.ambient().dimension()
+            h = PolynomialRing(ZZ,'h',1).gens()[0]
+            Segre_Class = sum([(-1)**(n-k-1) * sum([(-1)**i*binomial(n-k,i)*m**(n-k-i)*d[i] for i in range(n-k+1)]) * h**(n-k) for k in range(r+1)])
+            Chern_Fulton_Class = Segre_Class * (1+h)**(n+1)
+            self._topological_euler_characteristic = Chern_Fulton_Class.list()[n]
+            return self._topological_euler_characteristic
+
+    def _topological_Euler_characteristic_macaulay2(self, verbose=None):
+        r"""See :meth:`topological_euler_characteristic` for documentation."""
+        if verbose is None:
+            verbose = __VERBOSE__
+        try:
+            return self._topological_euler_characteristic
+        except AttributeError:
+            if verbose:
+                print("-- running Macaulay2 function eulerCharacteristic()... --")
+            X = macaulay2(self)
+            e = X.eulerCharacteristic().sage()
+            self._topological_euler_characteristic = e
+            if verbose:
+                print("-- function eulerCharacteristic() has terminated. --")
+            return self._topological_euler_characteristic
+
+    def point(self, verbose=None, UseMacaulay2=None):
```

Comment:
I think it would be better to use a keyword `algorithm`, with options `None`, `'macaulay2'`, `'sage'`

Reply:
I agree. I'll change it with the next commit.